RFC: Add a typing-compatible mechanism for ad-hoc attributes on various objects

bluetech opened the pull request:
pytest has several instances where plugins set their own attributes on objects they receive in hooks, like nodes and config. Since plugins are detached from these object's definition by design, this causes a problem for type checking because these attributes are not defined and mypy complains. And even if ignored, the values are untyped.

Fix this by giving these objects a "store" which can be used by plugins in a type-safe manner. See the docstring of the `Store` class in the new file `_pytest/store.py` for details.

Currently this mechanism is private. We can consider exposing it at a later point.

Comments welcome, including "too complicated" or ways to improve.

[Outdated] For the RFC, I only added a `Store` to `Node` and only converted the `skipping.py` plugin, which is sufficient to demonstrate the concept. If accepted, I will convert all of the other cases I can find.

bluetech commented:
#3740 might be related. I don't consider lifetimes etc. cc @RonnyPfannschmidt - if you'd like the expand on what you had in mind.

RonnyPfannschmidt commented:
starting with a plain store and creating a enabling api is actually a pretty good increment,

i believe the store tokens can be expanded with a boolean marker for what life-cycle they contribute in/to

i really like the token system and how it interacts wth the rest, 

instead of a boolean we might want to have a separate "scope bound" store and integrate a general concept of perishables in the stores (where the scope bound store is simply taken down a number of times and the normal store just once)

RonnyPfannschmidt commented:
i`d like to hear from the others as well - i think we should spend a little bit of time to build a mutual understanding of how to progress from this starting point to scopes and teadown of the storages /specific storage items

we might want to expand the lookup/addition, so that we cannot just push a value to a token, but also enter a context manager that yields a value to that type to express cleanup

to mark this api as "extra private" i propose adding _ to the starts of the Store/Token class names and to make it clear that the api is for internal experimentation and will do multiple breaking iterations

bluetech commented:
Made a few updates:

- Changed `StoreToken.mint(str)` to `StoreToken[str].mint()`.
- Added a store to `Config`.
- Converted a few more plugins.
- Added a simple test.

---

> instead of a boolean we might want to have a separate "scope bound" store and integrate a general concept of perishables in the stores (where the scope bound store is simply taken down a number of times and the normal store just once)

Hmm, interesting. Right now the plugins manage it manually. For example, if they store something during `pytest_configure`, the delete it during `pytest_unconfigure`. Is this what you had in mind? Or do you mean fixture-like scopes (function, class, module etc.)?

> we might want to expand the lookup/addition, so that we cannot just push a value to a token, but also enter a context manager that yields a value to that type to express cleanup

I wonder if this does not duplicate the functionality of finalizers which are present in a few places?

> to mark this api as "extra private" i propose adding _ to the starts of the Store/Token class names and to make it clear that the api is for internal experimentation and will do multiple breaking iterations

Probably a good idea, I'll change it at least before/if this is accepted.

bluetech commented:
Another update:
- Change "token" to "key". It's shorter and more understandable.
- Remove mint(), to forge a new key just use the constructor.

The-Compiler commented:
This seems a nice idea to me even regardless of typing - I always found it a bit weird to just assign random attributes to a config object everywhere.

RonnyPfannschmidt commented:
?

RonnyPfannschmidt commented:
@bluetech i have seen your note on one of the closed prs wrt the effort, i would like to see this one go in soon (to allow experimentation with more in top)

how do you want things to move forward for this in particular and for the full things in general

bluetech commented:
> how do you want things to move forward for this in particular and for the full things in general

I moved all of my pending typing work to a single branch on my repo to make my life easier. Overall, for me, the `Store` type is just a way to make typing happy, I'm not sure it makes sense to add the complexity if not for typing. That's why I closed this PR as well.

But if you think it makes sense to add regardless, I can reopen it. Hmm... but because I deleted the branch, github doesn't allow to reopen it, even though I recreated the branch. :confused:

RonnyPfannschmidt commented:
@bluetech it certainly adds a clean interface to build on top of, and it goes into a good direction for having the elements that enrich items/config decouple much better

i would like to have this as part of the next feature release so we can then do experiments for fixture internals and other details that should put information into more sensible places

(for me the typing details are not that interesting, but the explicit storage mechanism opens up new and interesting possibilities to explore)

bluetech commented:
OK, I guess I can live with one PR :) I'll resubmit it in a new PR since I can't reopen this one.
